sys/include/net/ieee802154_security.h

```diff
@@ -222,21 +253,204 @@ typedef enum {
     IEEE802154_SEC_SCF_KEYMODE_HW_INDEX         = 0x03  /**< Key is determined from 8 byte key source and key index */
 } ieee802154_sec_scf_keymode_t;
 
+/**
+ * @brief   Device addressing mode of device descriptor
+ */
+typedef enum {
+    IEEE802154_SEC_DEV_ADDRMODE_NONE            = 0x00, /**< PAN coordinator */
+    IEEE802154_SEC_DEV_ADDRMODE_SHORT           = 0x01, /**< Short address as device address */
+    IEEE802154_SEC_DEV_ADDRMODE_LONG            = 0x02  /**< Long address as device address */
+} ieee802154_sec_dev_addrmode_t;
+
 /**
  * @brief   IEEE 802.15.4 security error codes
  */
 typedef enum {
     IEEE802154_SEC_OK,                                  /**< Everything went fine */
     IEEE802154_SEC_FRAME_COUNTER_OVERFLOW,              /**< The requested operation would let the frame counter overflow */
+    IEEE802154_SEC_FRAME_COUNTER_ERROR,                 /**< The received frame counter is less than expected */
     IEEE802154_SEC_NO_KEY,                              /**< Could not find the key to perform a requested cipher operation */
+    IEEE802154_SEC_NO_DEV,                              /**< Could not find peer device to check the frame counter */
+    IEEE802154_SEC_NO_PEER,                             /**< Device was found but pairing has not been done to use the key */
     IEEE802154_SEC_MAC_CHECK_FAILURE,                   /**< The computet MAC did not match */
     IEEE802154_SEC_UNSUPORTED,                          /**< Unsupported operation */
 } ieee802154_sec_error_t;
 
+/**
+ *  @brief  Value to represent an invalid index
+ *
+ *  We shall support 255 valid key slots [0, 254].
+ */
+#define IEEE802154_SEC_NO_IDENT     (UINT8_MAX)
+
+/**
+ * @brief   Type to identify a key
+ */
+typedef uint8_t ieee802154_sec_key_descriptor_t;
+
+/**
+ * @brief   Type to identify a peer device
+ */
+typedef uint8_t ieee802154_sec_dev_descriptor_t;
+
+/**
+ * @brief   Type to identify a key lookup
+ */
+typedef uint8_t ieee802154_sec_key_lookup_descriptor_t;
+
+/**
+ * @brief   Stores peer device information
+ */
+typedef struct ieee802154_sec_dev_lookup {
+    /**
+     * @brief   PAN ID
+     */
+    uint8_t pan_id[2];
+    /**
+     * @brief   Short address
+     */
+    uint8_t short_addr[IEEE802154_SHORT_ADDRESS_LEN];
+    /**
+     * @brief   Extended address
+     */
+    uint8_t long_addr[IEEE802154_LONG_ADDRESS_LEN];
+    /**
+     * @brief   Incoming frame counter for replay protection
+     */
+    uint32_t fc;
+} ieee802154_sec_dev_lookup_t;
+
+/**
+ * @brief   Lookup table for peer devices to securely communicate with
```

Comment:
I didn't quite fully wrap my head around this, but could [`netstats_nb_t`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/netstats.h#L74) help here? (as in #16312)

Reply:
> Do I understand it correctly that with this we have a separate key for each pair of devices?

This would be the implicit key mode, which means after key exchange phase with a peer, the key to use is implicitly known by both, from the peer device address.
This requires 1 Keydescriptor + 1 KeyLookupDescriptor + 1 DeviceDescriptor per peer.

There are explicit keys, also called group keys. Let´s say a PAN coodinator is able to distribute some kind of key database in a network (or preconfigured).
The keys are somehow advertised with an indentifier (index). Each node will know `key at index 1` is `<KEY1>` and `key at index 2` is `<KEY2>` and so on.
So devices in that PAN use for example `key_mode = IEEE802154_SEC_SCF_KEYMODE_INDEX` and `kex_index = i` in the frame security header.
And if two peers are aware of the same key database, they will know which key is key `i`.
This requires 1 KeyDescriptor + 1 KeyLookupDescriptor + n DeviceDescriptors, for `n` peers.

We do need a DeviceDescriptor for each peer we communicate with, anyway to chack the incoming frame counter. 
I made a sketh in `sys/net/link_layer/ieee802154/security.c`. But I noticed there is one missing edge.

> So for a simple star topology CONFIG_IEEE802154_SEC_DEFAULT_DEVSTORE_SIZE would be 1 for the leaf nodes and n for the border router, where n is the number of leaf nodes?

Yes, given that communication always happens through the border router as a PAN coordinator. But I think RIOT is not exposing the concept of a PAN coordinator at the moment.

> I didn't quite fully wrap my head around this, but could netstats_nb_t help here? (as in #16312)

I would not extend `netstats_nb_t` because this is volatile information but `ieee802154_sec_dev_lookup_t` must go into persistent storage.
The DeviceDescriptor and especially the frame counter in it must remain, even if a neighbor is thrown out of the neighbor cache.
I think the neighbor cache could be useful to bootstrap peering.
In the manner of: "If I see a neighbor that I don´t have a security DeviceDescrptor for, I start some kind of peering phase and key negotiation."